File: 프로그래머스/LV3/코딩테스트 연습/월간 코드 챌린지 시즌1/스타 수열/스타 수열.py
def solution(a):
    dic = {n : [0, -1] for n in list(set(a))}
    c = 0

    for i in range(len(a)):
        if i > 0 and dic[a[i]][1] != i-1 and a[i-1] != a[i]:
            dic[a[i]][0] += 1
            dic[a[i]][1] = i-1
        elif i < len(a)-1 and a[i] != a[i+1]:
            dic[a[i]][0] += 1
            dic[a[i]][1] = i+1
        c = max(c, dic[a[i]][0])

    return c * 2 if len(a) >= c * 2 else (len(a) - c) * 2

File: 프로그래머스/LV3/코딩테스트 연습/월간 코드 챌린지 시즌1/스타 수열/스타 수열.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
using namespace std;

int solution(std::vector<int> a) {
    unordered_map<int, pair<int, int>> map;
    unordered_set<int> s(a.begin(), a.end());
    int c = 0;

    for(auto n : s)
        map.insert({n, make_pair(0, -1)});
    
    for(int i = 0; i < a.size(); i++) {
        if(i > 0 && map[a[i]].second != i-1 && a[i-1] != a[i]) {
            map[a[i]].first++;
            map[a[i]].second = i-1;
        } 
        else if(i < a.size()-1 && a[i] != a[i+1]) {
            map[a[i]].first++;
            map[a[i]].second = i+1;            
        }
        c = max(c, map[a[i]].first);
    }

    return a.size() >= c * 2 ? c * 2 : (a.size()-c) * 2;
}